File: baekjoon/14888.py
import copy
n = int(input())    # 수의 개수
numbers = list(map(int, input().split()))  # 수
p, m, mul, div = map(int, input().split())

MAX = -1e10
MIN = 1e10
def dfs(now, ret):
    global MAX, MIN, p, m, mul, div
    if now == n:
        MAX = max(MAX, ret)
        MIN = min(MIN, ret)
        return
    else:
        if p>0:
            p -= 1
            dfs(now+1, ret+numbers[now])
            p += 1
        if m>0:
            m -= 1
            dfs(now+1, ret-numbers[now])
            m += 1
        if mul > 0:
            mul -= 1
            dfs(now+1, ret * numbers[now])
            mul += 1
        if div > 0:
            div -= 1
            if ret < 0:
                dfs(now+1, -1* (abs(ret)//numbers[now]))
            else:
                dfs(now+1, ret//numbers[now])
            div += 1


dfs(1, numbers[0])
print(MAX)
print(MIN)

File: baekjoon/14888.js
// https://www.acmicpc.net/problem/14888
// 연산자 끼워넣기
// 브루트포스, 백트래킹

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const [n, input_1, input_2] = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n");
const numArr = input_1.trim().split(" ").map(Number);
let [plus, minus, mul, div] = input_2.trim().split(" ").map(Number);

let MIN = 1e9;
let MAX = -1e9;
function solution(now, ret) {
  if (now == parseInt(n) - 1) {
    MIN = Math.min(MIN, ret);
    MAX = Math.max(MAX, ret);
    return;
  } else {
    if (plus > 0) {
      plus--;
      solution(now + 1, ret + numArr[now + 1]);
      plus++;
    }
    if (minus > 0) {
      minus--;
      solution(now + 1, ret - numArr[now + 1]);
      minus++;
    }
    if (mul > 0) {
      mul--;
      solution(now + 1, ret * numArr[now + 1]);
      mul++;
    }
    if (div > 0) {
      div--;
      ret < 0
        ? solution(now + 1, Math.floor((ret * -1) / numArr[now + 1]) * -1)
        : solution(now + 1, Math.floor(ret / numArr[now + 1]));
      div++;
    }
  }
}

solution(0, numArr[0]);
console.log(`${MAX}\n${MIN}`);
